File: tests/algorithms/x/Pascal/data_structures/binary_tree/number_of_possible_binary_trees.pas
{$mode objfpc}
program Main;
uses SysUtils;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
function _input(): string;
var s: string;
begin
  if EOF(Input) then s := '' else ReadLn(s);
  _input := s;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  input_str: string;
  node_count: integer;
  bst: integer;
  bt: integer;
  k: integer;
  n: integer;
function binomial_coefficient(n: integer; k: integer): integer; forward;
function catalan_number(node_count: integer): integer; forward;
function factorial(n: integer): integer; forward;
function binary_tree_count(node_count: integer): integer; forward;
function binomial_coefficient(n: integer; k: integer): integer;
var
  binomial_coefficient_result_: integer;
  binomial_coefficient_kk: integer;
  binomial_coefficient_i: int64;
begin
  binomial_coefficient_result_ := 1;
  binomial_coefficient_kk := k;
  if k > (n - k) then begin
  binomial_coefficient_kk := n - k;
end;
  for binomial_coefficient_i := 0 to (binomial_coefficient_kk - 1) do begin
  binomial_coefficient_result_ := binomial_coefficient_result_ * (n - binomial_coefficient_i);
  binomial_coefficient_result_ := binomial_coefficient_result_ div (binomial_coefficient_i + 1);
end;
  exit(binomial_coefficient_result_);
end;
function catalan_number(node_count: integer): integer;
begin
  exit(binomial_coefficient(2 * node_count, node_count) div (node_count + 1));
end;
function factorial(n: integer): integer;
var
  factorial_result_: integer;
  factorial_i: int64;
begin
  if n < 0 then begin
  writeln('factorial() not defined for negative values');
  exit(0);
end;
  factorial_result_ := 1;
  for factorial_i := 1 to (n + 1 - 1) do begin
  factorial_result_ := factorial_result_ * factorial_i;
end;
  exit(factorial_result_);
end;
function binary_tree_count(node_count: integer): integer;
begin
  exit(catalan_number(node_count) * factorial(node_count));
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  writeln('Enter the number of nodes:');
  input_str := _input();
  node_count := StrToInt(input_str);
  if node_count <= 0 then begin
  writeln('We need some nodes to work with.');
end else begin
  bst := catalan_number(node_count);
  bt := binary_tree_count(node_count);
  writeln('Given', ' ', node_count, ' ', 'nodes, there are', ' ', bt, ' ', 'binary trees and', ' ', bst, ' ', 'binary search trees.');
end;
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.

File: tests/algorithms/x/PHP/data_structures/binary_tree/number_of_possible_binary_trees.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _intdiv($a, $b) {
    if ($b === 0 || $b === '0') {
        throw new DivisionByZeroError();
    }
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv(intval($a), intval($b));
}
$__start_mem = memory_get_usage();
$__start = _now();
  function binomial_coefficient($n, $k) {
  global $bst, $bt, $input_str, $node_count;
  $result = 1;
  $kk = $k;
  if ($k > $n - $k) {
  $kk = $n - $k;
}
  for ($i = 0; $i < $kk; $i++) {
  $result = $result * ($n - $i);
  $result = _intdiv($result, ($i + 1));
};
  return $result;
};
  function catalan_number($node_count) {
  global $bst, $bt, $input_str;
  return binomial_coefficient(2 * $node_count, $node_count) / ($node_count + 1);
};
  function factorial($n) {
  global $bst, $bt, $input_str, $node_count;
  if ($n < 0) {
  echo rtrim('factorial() not defined for negative values'), PHP_EOL;
  return 0;
}
  $result = 1;
  for ($i = 1; $i < ($n + 1); $i++) {
  $result = $result * $i;
};
  return $result;
};
  function binary_tree_count($node_count) {
  global $bst, $bt, $input_str;
  return catalan_number($node_count) * factorial($node_count);
};
  echo rtrim('Enter the number of nodes:'), PHP_EOL;
  $input_str = trim(fgets(STDIN));
  $node_count = intval($input_str);
  if ($node_count <= 0) {
  echo rtrim('We need some nodes to work with.'), PHP_EOL;
} else {
  $bst = catalan_number($node_count);
  $bt = binary_tree_count($node_count);
  echo rtrim('Given') . " " . rtrim(json_encode($node_count, 1344)) . " " . rtrim('nodes, there are') . " " . rtrim(json_encode($bt, 1344)) . " " . rtrim('binary trees and') . " " . rtrim(json_encode($bst, 1344)) . " " . rtrim('binary search trees.'), PHP_EOL;
}
$__end = _now();
$__end_mem = memory_get_peak_usage(true);
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
